### PR TITLE
Dharvey ignite 2.5

### PR DESCRIPTION
**Pull request by DaveWHarvey**
Fix for IGNITE-9026.   I haven't tried this yet.  Though it would be good to get any feedback first.

### CODE REVIEW

**Comment by markusgay** [comment]
Hello,
The code changes look fine. Please make sure the Unit test org.apache.ignite.internal.managers.deployment.GridDeploymentMessageCountSelfTest.testTaskDeployment() succeeds with your changes.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java [hunk 1]
```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
                 GridIoPolicy.P2P_POOL);
         }
     }
-
+    
+    /**
+     * Queries whether a node is a direct or indirectly responsible for
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: Returns true if the given node is directly or indirectly responsible for the current {@linkplain GridDeploymentRequest request} (which is not an {@linkplain GridDeploymentRequest#isUndeploy undeploy) being processed by this thread.

@pre node != null

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java [hunk 2]
```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
                 GridIoPolicy.P2P_POOL);
         }
     }
-
+    
+    /**
+     * Queries whether a node is a direct or indirectly responsible for 
+     * the request being processed by this thread.
+     * 
+     * @param  node The node in question
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: Remove @param and @return as they are redundant to the above contract.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java [hunk 3]
```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
...
+     * @return true if the node is active
+     */
+    boolean nodeOriginatedCurrentRequest(final ClusterNode node) {
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: Why do we need this to be final?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java [hunk 4]
```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
...
+       // activeReqNodeIds is thread local.
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: assert node != null

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java [hunk 5]
```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
...
      *      be obsolete.
+     * @param exclusionList - Nodes to avoid recursively searching, in addition to 
+     *      the node(s) that forwarded this request initially.
      * @return Either response value or {@code null} if timeout occurred.
      * @throws IgniteCheckedException Thrown if there is no connection with remote node.
      */
     @SuppressWarnings({"SynchronizationOnLocalVariableOrMethodParameter"})
     GridDeploymentResponse sendResourceRequest(final String rsrcName, IgniteUuid clsLdrId,
-        final ClusterNode dstNode, long threshold) throws IgniteCheckedException {
+        final ClusterNode dstNode, long threshold, final Collection<UUID> nodesToSkip) throws IgniteCheckedException {
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: Why final? Is this an explicit convention in the Ignite code?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java [hunk 6]
```diff
@@ -337,50 +337,63 @@ void sendUndeployRequest(String rsrcName, Collection<ClusterNode> rmtNodes) thro
...
+     * no response until threshold time, method returns null.  The receiver
+     * may forward the request to other nodes, but not to nodes that have been
+     * searched already, or nodes we intend to search directly.
      *
+     * The exclusion list severely weakens the potential n-squared behavior when a 
+     * resource does not exist. If the resource does not exist, all nodes 
+     * in the graph must be touched at least once.   To avoid all multiple touches 
+     * for the resource-not-found case, the response would need to include the 
+     * nodes forwarded to, so they could be added to the exclusion list.  However
+     * that seems like overkill for this kinds of graphs peer class loading will generate.
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: that seems like overkill for the kinds of graphs peer class loading will generate.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentClassLoader.java [hunk 7]
```diff
@@ -596,9 +596,16 @@ private GridByteArrayList sendClassRequest(String name, String path) throws Clas
 
                 continue;
             }
+            
+            // Must not send to the participants that asked us for the resource.
+            // NOTE: We could probably assert this, since this should not be called in the
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentClassLoader.java]: The note here is a but ambiguous. Do you mean in the context of a resource request - i.e., a GridDeploymentRequest message?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentClassLoader.java [hunk 8]
```diff
@@ -615,8 +622,20 @@ else if (log.isDebugEnabled())
                     continue;
                 }
 
-                if (res.success())
+                if (res.success()) {
+                   
+                    // If we searched other nodes w/o succeeding first, then 
+                    // make the node we found the resource on the first to search next time.
```
Comment (mfagnani-work) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentClassLoader.java]: I would consider replacing "node" with "participant" in all the comments you've added as you did that above and it's a much clearer description than "node"

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java (hunk 2 above)
Comment (DaveWHarvey) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: Looking at their coding  style, it seems that they require @param and @return (but it is not important that they add value to the reader).     I'm willing to add stuff that they may not want, but I'm going to try to stick with their convention by specifying all of these.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java (hunk 5 above)
Comment (DaveWHarvey) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/deployment/GridDeploymentCommunication.java]: Seems pretty inconsistent.

**Pull request merged**